### PR TITLE
MOL-290/ Fix 2nd channel 500 issues

### PR DESCRIPTION
**Pull request by marekrzytki**
Add mollie based cart context to retrieve order when request comes from Mollie

### CODE REVIEW

File: src/Context/Cart/MollieResourceBasedCartContext.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+use Symfony\Component\HttpFoundation\RequestStack;
+
+final class MollieResourceBasedCartContext implements CartContextInterface
+{
+    private RequestStack $requestStack;
+
+    private MollieResourceOrderResolverInterface $mollieResourceOrderResolver;
+
+    public function __construct(
+        RequestStack $requestStack,
```
Comment (senghe) [inline src/Context/Cart/MollieResourceBasedCartContext.php:28]: This dependency makes the service usable only in HTTP requests. I thought you should know ;)

File: src/Context/Cart/MollieResourceBasedCartContext.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        MollieResourceOrderResolverInterface $mollieResourceOrderResolver
+    ) {
+        $this->requestStack = $requestStack;
+        $this->mollieResourceOrderResolver = $mollieResourceOrderResolver;
+    }
+
+    public function getCart(): OrderInterface
+    {
+        $request = $this->getMasterRequest();
+        $mollieResourceId = $request->get('id');
```
Comment (senghe) [inline src/Context/Cart/MollieResourceBasedCartContext.php:38]: How about casting it to int?

File: src/Context/Cart/MollieResourceBasedCartContext.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            ;
+
+        return $order;
+    }
+
+    private function getMasterRequest(): Request
+    {
+        $masterRequest = $this->requestStack->getMasterRequest();
+        if (null === $masterRequest) {
+            throw new CartNotFoundException('There is no master request on request stack.');
```
Comment (senghe) [inline src/Context/Cart/MollieResourceBasedCartContext.php:56]: Is there any chance to occur?

File: src/Resolver/Order/MollieResourceOrderResolver.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                                $mollieResourceId
+                            )
+                        );
+
+                    throw new ResourceNotFoundException();
+            }
+        } catch (ApiException $e) {
+            $this->loggerAction->addNegativeLog(sprintf('API call failed: %s', htmlspecialchars($e->getMessage())));
+
+            throw new \Exception(sprintf('API call failed: %s', htmlspecialchars($e->getMessage())));
```
Comment (senghe) [inline src/Resolver/Order/MollieResourceOrderResolver.php:66]: I would rethrow $e

File: spec/Context/Cart/MollieResourceBasedCartContextSpec.php [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+        $mollieResourceOrderResolver
+            ->resolve($mollieResourceId)
+            ->willReturn($order);
+
+        $this->getCart()
+            ->shouldReturn($order);
+    }
+
+    public function it_throws_exception_if_there_is_no_mollie_resource_id(
```
Comment (senghe) [inline spec/Context/Cart/MollieResourceBasedCartContextSpec.php:63]: Where is the scenario of missing master request? 🙃

File: src/Resources/config/services/context.xml [hunk 6]
```diff
@@ -7,5 +7,14 @@
...
             <argument type="service" id="security.token_storage"/>
         </service>
+
+        <service id="bitbag_sylius_mollie_plugin.context.mollie_resource_based_cart_context"
+                class="BitBag\SyliusMolliePlugin\Context\Cart\MollieResourceBasedCartContext"
+        >
+            <argument type="service" id="request_stack"/>
+            <argument type="service" id="bitbag_sylius_mollie_plugin.resolver.order.mollie_resource_order_resolver"/>
+
+            <tag name="sylius.context.cart" priority="-1000" />
```
Comment (senghe) [inline src/Resources/config/services/context.xml:17]: Why priority -1000?

**Review by senghe — CHANGES_REQUESTED** [review]

File: src/Context/Cart/MollieResourceBasedCartContext.php (hunk 2 above)
Comment (marekrzytki) [reply src/Context/Cart/MollieResourceBasedCartContext.php:38]: I can't as mollie resource id is made up with letters and digits

File: src/Resources/config/services/context.xml (hunk 6 above)
Comment (marekrzytki) [reply src/Resources/config/services/context.xml:17]: This context has small scope of usage, so I want it to be last in queue

File: src/Context/Cart/MollieResourceBasedCartContext.php (hunk 3 above)
Comment (marekrzytki) [reply src/Context/Cart/MollieResourceBasedCartContext.php:56]: There is no chance, I will remove it.

File: src/Resolver/Order/MollieResourceOrderResolver.php (hunk 4 above)
Comment (marekrzytki) [reply src/Resolver/Order/MollieResourceOrderResolver.php]: Done

File: spec/Context/Cart/MollieResourceBasedCartContextSpec.php (hunk 5 above)
Comment (marekrzytki) [reply spec/Context/Cart/MollieResourceBasedCartContextSpec.php:63]: This part is removed

File: src/Resolver/Order/MollieResourceOrderResolver.php [hunk 7]
```diff
@@ -63,7 +63,7 @@ public function resolve(string $mollieResourceId): OrderInterface
         } catch (ApiException $e) {
             $this->loggerAction->addNegativeLog(sprintf('API call failed: %s', htmlspecialchars($e->getMessage())));
 
-            throw new \Exception(sprintf('API call failed: %s', htmlspecialchars($e->getMessage())));
+            throw new ApiException($e->getMessage());
```
Comment (senghe) [inline src/Resolver/Order/MollieResourceOrderResolver.php:66]:
```suggestion
            throw $e;
```

**Review by senghe — CHANGES_REQUESTED** [review]

**Review by senghe — APPROVED** [review]

**Pull request merged**